cache/src/main/scala/coursier/Cache.scala

```diff
@@ -946,50 +947,84 @@ object Cache {
+    }
+    else {
 
-    implicit val pool0 = pool
-
-    val checksums0 = if (checksums.isEmpty) Seq(None) else checksums
-
-    val res = EitherT {
-      download(
-        artifact,
-        cache,
-        checksums = checksums0.collect { case Some(c) => c }.toSet,
-        cachePolicy,
-        pool,
-        logger0 = logger,
-        ttl = ttl
-      ).map { results =>
-        val checksum = checksums0.find {
-          case None => true
-          case Some(c) =>
-            artifact.checksumUrls.get(c).exists { cUrl =>
-              results.exists { case ((_, u), b) =>
-                u == cUrl && b.isRight
+      implicit val pool0 = pool
+      val checksums0: Seq[Option[String]] = if (checksums.isEmpty) Seq(None) else checksums
+
+      val res: EitherT[Task, FileError, (File, Option[String])] = EitherT {
+        download(
+          artifact,
+          cache,
+          checksums = checksums0.collect { case Some(c) => c }.toSet,
+          cachePolicy,
+          pool,
+          logger0 = logger,
+          ttl = ttl
+        ).map { results =>
+          val checksum = checksums0.find {
+            case None => true
+            case Some(c) =>
+              artifact.checksumUrls.get(c).exists { cUrl =>
+                results.exists { case ((_, u), b) =>
+                  u == cUrl && b.isRight
+                }
               }
-            }
-        }
+          }
 
-        val ((f, _), res) = results.head
-        res.right.flatMap { _ =>
-          checksum match {
-            case None =>
-              // FIXME All the checksums should be in the error, possibly with their URLs
-              //       from artifact.checksumUrls
-              Left(FileError.ChecksumNotFound(checksums0.last.get, ""))
-            case Some(c) => Right((f, c))
+          val ((f, _), res) = results.head
+          res.right.flatMap { _ =>
+            checksum match {
+              case None =>
+                // FIXME All the checksums should be in the error, possibly with their URLs
+                //       from artifact.checksumUrls
+                Left(FileError.ChecksumNotFound(checksums0.last.get, ""))
+              case Some(c) => Right((f, c))
+            }
           }
         }
       }
-    }
 
-    res.flatMap {
-      case (f, None) => EitherT(Task.now[Either[FileError, File]](Right(f)))
-      case (f, Some(c)) =>
-        validateChecksum(artifact, c, cache, pool).map(_ => f)
+      res.flatMap {
+        case (file0, None) => EitherT(Task.now[Either[FileError, File]](Right(file0)))
+        case (file0, Some(shaType)) =>
+          validateChecksum(artifact, shaType, cache, pool).map(_ => file0)
+      }.leftFlatMap {
+        case _: FileError.WrongChecksum =>
+          val badFile = localFile(artifact.url, cache, artifact.authentication.map(_.user))
+          badFile.delete()
+          Console.err.println(s"Bad file deleted: ${badFile.getAbsolutePath}")
```

Comment:
Use `logger` instead? (a method can be added to it)

Reply:
Added